test/e2e/node/batchresource.go

```diff
@@ -0,0 +1,184 @@
+			} else if errors.IsNotFound(err) {
+				framework.Logf("cannot get slo-controller-config %s/%s, try to create a new one",
+					sloConfigNamespace, sloConfigName)
+			} else {
+				framework.Failf("failed to get slo-controller-config %s/%s, got unexpected error: %v",
+					sloConfigNamespace, sloConfigName, err)
+			}
+
+			// If configmap is created, try to patch it with colocation enabled.
+			// If not exist, create the slo-controller-config.
+			// NOTE: slo-controller-config should not be modified by the others during the e2e test.
+			ginkgo.By("Prepare slo-controller-config to enable colocation")
+			if isConfigCreated {
+				needUpdate := false
+				rollbackData := map[string]string{}
+				if configMap.Data == nil {
+					needUpdate = true
+				} else if configMap.Data[apiext.ColocationConfigKey] != colocationEnabledConfigData {
+					rollbackData[apiext.ColocationConfigKey] = configMap.Data[apiext.ColocationConfigKey]
+					needUpdate = true
+				}
+
+				if _, ok := rollbackData[apiext.ColocationConfigKey]; ok && needUpdate {
+					defer rollbackSLOConfigData(f, rollbackData)
+				}
+
+				if needUpdate {
+					framework.Logf("colocation is not enabled in slo-controller-config, need update")
+					newConfigMap := configMap.DeepCopy()
+					newConfigMap.Data[apiext.ColocationConfigKey] = colocationEnabledConfigData
+					newConfigMapUpdated, err := c.CoreV1().ConfigMaps(sloConfigNamespace).Update(context.TODO(), newConfigMap, metav1.UpdateOptions{})
+					framework.ExpectNoError(err)
+					framework.Logf("update slo-controller-config successfully, data: %v", newConfigMapUpdated.Data)
+					configMap = newConfigMapUpdated
+				} else {
+					framework.Logf("colocation is already enabled in slo-controller-config, keep the same")
+				}
+			} else {
+				framework.Logf("slo-controller-config does not exist, need create")
+				newConfigMap, err := manifest.ConfigMapFromManifest("node/slo-controller-config.yaml")
+				framework.ExpectNoError(err)
+				newConfigMapCreated, err := c.CoreV1().ConfigMaps(sloConfigNamespace).Create(context.TODO(), newConfigMap, metav1.CreateOptions{})
+				framework.ExpectNoError(err)
+				framework.Logf("create slo-controller-config successfully, data: %v", newConfigMapCreated.Data)
+				configMap = newConfigMapCreated
+
+				defer rollbackSLOConfigObject(f)
+			}
+
+			ginkgo.By("Check node allocatable for batch resources")
+			totalCount, allocatableCount := len(nodeList.Items), 0 // assert totalCount > 0
+			gomega.Eventually(func() bool {
+				for i := range nodeList.Items {
+					node := &nodeList.Items[i]
+					isAllocatable, msg := isNodeHasBatchResourcesAllocatable(node)
+					if isAllocatable {
+						allocatableCount++
+					} else {
+						framework.Logf("node %s has no batch resources allocatable, msg: %s", node.Name, msg)
+					}
+				}
+				if float64(allocatableCount) > float64(totalCount)*minNodesBatchResourceAllocatableRatio {
```

Comment:
can we also check the exactly batch total resource by mocking NodeMetric CR of node?

Reply:
Yes, but a little complicated.
By default, the test case runs when the koordinator is deployed in a general configuration. Before mocking a NodeMetric CR, we should disable the node metric reporting by mutating the spec of the koordlet DaemonSet.

Reply:
We may not mock a NodeMetric CR. Instead, we get the latest NodeMetric CR and roughly verify the node allocatable batch resources.

Reply:
the later one sound good, since NodeMetric won't update frequently.

Reply:
Done. Now we estimate the lower bound of the batch resources using `batch.total >= node.total - node.total * cpuReclaimRatio - nodeMetric.usage - node.total * maxDiffRatio`.